nixos/geoclue2: make configurable, can whitelist applications

worldofpeace opened the pull request:
###### Motivation for this change
Non-functioning [elementary geoclue2 agent](https://github.com/elementary/pantheon-agent-geoclue2)

Eventually we'd like something similar to [this](https://gitlab.freedesktop.org/geoclue/geoclue/issues/74) but I'm unsure also, as mentioned, if elementary should be adding their agent into `geoclue.conf` upstream.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---
cc @jtojnar

jtojnar commented on nixos/modules/services/desktops/geoclue2.nix:
```diff
@@ -61,6 +88,106 @@ in
     };
+
+    environment.etc."geoclue/geoclue.conf".text = ''
+    # Configuration file for Geoclue
```
Do we need most of these options or can we leave them implicit?

worldofpeace replied:
I'm seeing other projects just leaving it as is aside from using a custom geolocation api key from google.

jtojnar replied:
One concern with the explicit defaults is that user cannot override them. If they cannot be omitted, we would need to make them configurable.

worldofpeace replied:
Very true, maybe I'll make an option only to create more application configurations (since that's the only thing I'll need currently), and remove `extraConfig`. I'll deal with the rest later.
Though I'm not sure of additional useful geoclue configuration.

jtojnar replied:
It is mainly about the things already in the text. For example, someone might want to switch to google services or other provider. Or change their leaderboard name.

Infinisil replied:
Yeah I also think this config file should be as minimal as possible, because it's not really possible to change anything in it, only add to it. The default upstream file is intended to be modified, which we can't like this

Alternatively, create a `config` option that transforms a Nix value into the config file syntax, and provide upstream defaults via `config.gnome-color-panel.allowed = mkDefault true;`. Then it's possible to override everything. I'm doing pretty much the same for https://github.com/NixOS/nixpkgs/pull/41467 (which I still need to finish)

jtojnar commented on nixos/modules/services/desktops/geoclue2.nix:
```diff
@@ -35,23 +37,48 @@ in
+        type = types.lines;
+        default = "";
+        example = ''
+        [redshift]
```
This should be indented

jtojnar commented on nixos/modules/services/desktops/geoclue2.nix:
```diff
@@ -61,6 +186,104 @@ in
+
+    # Whitelist of desktop IDs (without .desktop part) of all agents we recognise,
+    # separated by a ';'.
+    whitelist=${optionalString cfg.enableDemoAgent semiColon "geoclue-demo-agent"}${semiColon "gnome-shell"}${semiColon cfg.extraWhitelistedAgents}
```
This should not work for many reasons. I would try something like `whitelist=${let whitelist = optional cfg.enableDemoAgent "geoclue-demo-agent" ++ singleton "gnome-shell" ++ cfg.extraWhitelistedAgents; in semiColon whitelist}`

michaelpj commented on nixos/modules/services/desktops/geoclue2.nix:
```diff
@@ -61,6 +186,104 @@ in
+    system=true
+    users=
+
+    [org.gnome.Shell]
```
Could we not move these to the gnome module now?

worldofpeace replied:
These are the defaults upstream. Though I feel the use of geoclue2 implies gnome does it not?

jtojnar replied:
Nope, geoclue is an independent freedesktop component. Also we do not have gnome-flashback module yet, so gnome-panel sections would have nowhere to go.

michaelpj commented on nixos/modules/services/desktops/geoclue2.nix:
```diff
@@ -61,6 +186,104 @@ in
+    system=false
+    users=
+
+    [firefox]
```
Could these be in the default value of `extraAppConfig`, so they should just get merged in normally?

worldofpeace replied:
Sure, though that might make maintenance unclear

Infinisil commented on nixos/modules/services/desktops/geoclue2.nix:
```diff
@@ -61,6 +186,104 @@ in
+    users=
+
+    ${concatStringsSep "\n" (
+        map appConfigStr (
```
There's `concatMapStringsSep`

Infinisil commented on nixos/modules/services/desktops/geoclue2.nix:
```diff
@@ -7,6 +7,56 @@ with lib;
+
+  cfg = config.services.geoclue2;
+
+  tf = value: if value then "true" else "false";
```
There's `lib.boolToString`

Infinisil commented on nixos/modules/services/desktops/geoclue2.nix:
```diff
@@ -35,23 +85,98 @@ in
+
+      extraWhitelistedAgents = mkOption {
+        type = types.listOf types.str;
+        default =  [ "" ];
```
I think this should be `[]` by default.

Infinisil replied:
`[]` and `[ "" ]` are both lists of strings, but I don't see a reason to include `""` by default

jtojnar replied:
I agree, I think this will result in the `whitelist` field always containing an empty string due to the merge strategy.

michaelpj commented:
This is still relevant - upstream have made some progress in making "system" apps not need the agent, so once they do a release we'll want to switch to whitelisting using an approach like this.

worldofpeace commented:
@michaelpj Cool, I'll finish cleaning this up in that case.

worldofpeace commented:
So I'm ok with this, though it's rather unfortunate that I wrote this without awareness of `lib.generators.toINI`.
That would have made this much simpler.
If any maintainer feels like making that change feel free to push it.

@michaelpj 
> Note that in recent versions they've dropped the agent requirement for "system" apps, so we can potentially drop that whole annoyance and just go back to modifying the config file.

I think that means there's no usecase for `enableDemoAgent` or  `extraWhitelistedAgents`, and the hacks for redshift can be dropped and replaced with
```
services.geoclue2.appConfig."redshift" = {
  isAllowed = true;
  isSystem = true;
};
```
I think...

michaelpj commented:
Yes, I agree. I'm happy to make that change once this is in, or you can do it.

worldofpeace committed:
nixos/geoclue2: make configurable, can whitelist applications

All options within geoclue.conf[1] have been made configurable.

Additonally, we can now specify whether or not GeoClue
should ask the agent to authorize an application like so:
```
services.geoclue2.appConfig."redshift" = {
  isAllowed = true;
  isSystem = true;
};
```

[1]: https://gitlab.freedesktop.org/geoclue/geoclue/blob/master/data/geoclue.conf.in

worldofpeace commented:
I've dropped `extraWhitelistedAgents` because people should just upstream that config to geoclue IIRC.

I feel like the removal of `enableDemoAgent` should be in a different pr, coinciding with the redshift change once this is in.

@jtojnar Is this ok, or do you still have reservations about it?

michaelpj commented:
I had a brief look at what they actually did to implement "systme apps don't need authorization" and I'm not sure that it corresponds to the `system` config option at all. I'll investigate separately.

michaelpj commented:
Things I would like to see from this:
- Use `toINI`
- Move whichever bits of configuration we can to the right modules. Just because upstream doesn't have a drop-in mechanism and so has to hardcode things doesn't mean that we have to.

worldofpeace commented:
> Move whichever bits of configuration we can to the right modules.

I can do that for gnome, however I can't  for `epiphany`, and for `firefox` I'm not sure.

worldofpeace committed:
nixos/geoclue2: modularize declared configurations

leotaku commented:
Is there anything we can do to push this forward? 

@michaelpj are you still interested in rewriting/seeing this rewritten to use `toINI`?

michaelpj commented:
I still think this would be useful indeed! I kind of gave up on geoclue, but it would be nice if it worked better.

worldofpeace commented:
Completed in https://github.com/NixOS/nixpkgs/pull/60681 :sparkles:
